modules/apps/frontend-js/frontend-js-bundle-config-extender/src/main/java/com/liferay/frontend/js/bundle/config/extender/internal/JSBundleConfigRegistry.java

```diff
@@ -53,46 +52,35 @@ public ServiceReference<ServletContext> addingService(
 				ServletContext servletContext = _bundleContext.getService(
 					serviceReference);
 
-				_jsConfigs.put(
-					serviceReference, new JSConfig(servletContext, url));
-
 				_lastModified = System.currentTimeMillis();
 
-				return serviceReference;
+				return new JSConfig(servletContext, url);
 			}
 		}
 
 		return null;
 	}
 
 	public Collection<JSConfig> getJSConfigs() {
-		return _jsConfigs.values();
+		return _serviceTrackerMap.values();
 	}
 
 	public long getLastModified() {
 		return _lastModified;
 	}
 
-	public long getTrackingCount() {
-		return _serviceTracker.getTrackingCount();
-	}
-
 	@Override
 	public void modifiedService(
-		ServiceReference<ServletContext> serviceReference,
-		ServiceReference<ServletContext> trackedServiceReference) {
+		ServiceReference<ServletContext> serviceReference, JSConfig jsConfig) {
 
-		removedService(serviceReference, trackedServiceReference);
+		removedService(serviceReference, jsConfig);
```

Comment:
@MrJohn1911 There's a trap here.

Before your change, the service tracker saves the constructed `JSConfig` object into a separate map, so when the service is modified, calling `removedService` will remove this `ServiceReference - JSConfig` mapping, and then `addingService` will re-add this mapping.

However, `ServiceTrackerMap` works differently. The code in `modifiedService` can not affect the object stored in the service tracker map. Actually, looking at the `ObjectServiceTrackerMapTest`, it will only update the key(s) upon service modification.

The `JSConfig` is just a POJO containing a bundle's `Liferay-JS-Config` URL and `ServletContext`.

We need to better understand whether we really need this `modifiedService` logic -- will `ServletContext` in Liferay really get modified? On the other hand, do we really need a `ServiceTrackerMap` instead of a `ServiceTrackerList`?

cc @tinatian